src/flamegraph/attrs.rs

```diff
@@ -0,0 +1,248 @@
+                                "class" => funcattrs.g.class = Some(value),
+                                "style" => funcattrs.g.style = Some(value),
+                                "onmouseover" => funcattrs.g.onmouseover = Some(value),
+                                "onmouseout" => funcattrs.g.onmouseout = Some(value),
+                                "onclick" => funcattrs.g.onclick = Some(value),
+                                "href" => funcattrs.a.href = Some(value),
+                                "target" => funcattrs.a.target = Some(value),
+                                "g_extra" => parse_extra_attrs(&mut funcattrs.g.extra, &value),
+                                "a_extra" => parse_extra_attrs(&mut funcattrs.a.extra, &value),
+                                _ => warn!("invalid attribute {} found for {}", name, func),
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(funcattr_map)
+    }
+
+    /// Return FrameAttrs for the given function name if it exists
+    pub fn frameattrs_for_func(&self, func: &str) -> Option<&FrameAttrs> {
+        self.0.get(func)
+    }
+}
+
+/// Attributes to set on the SVG elements of a frame
+#[derive(PartialEq, Eq, Debug, Default)]
+pub struct FrameAttrs {
+    /// The text to include in the `title` element.
+    /// If set to None, the title is dynamically generated based on the function name.
+    pub title: Option<String>,
+
+    pub g: GElementAttrs,
+    pub a: AElementAttrs,
+}
+
+/// Attributes to set on the SVG `g` element.
+/// Any of them set to `None` will get the default value.
+#[derive(PartialEq, Eq, Debug, Default)]
+pub struct GElementAttrs {
+    /// Defaults to "func_g"
+    pub class: Option<String>,
+
+    /// Will not be included if None
+    pub style: Option<String>,
+
+    /// Defaults to "s(this)"
+    pub onmouseover: Option<String>,
+
+    /// Defaults to "c()"
+    pub onmouseout: Option<String>,
+
+    /// Defaults to "zoom(this)"
+    pub onclick: Option<String>,
+
+    /// Extra attributes to include
+    pub extra: HashMap<String, String>,
+}
+
+/// Attributes to set on the SVG `a` element
+#[derive(PartialEq, Eq, Debug, Default)]
+pub struct AElementAttrs {
+    /// If set to None the `a` tag will not be added
+    pub href: Option<String>,
+    pub target: Option<String>,
+
+    /// Extra attributes to include
+    pub extra: HashMap<String, String>,
+}
+
+fn parse_extra_attrs(attrs: &mut HashMap<String, String>, s: &str) {
+    for (name, value) in name_value_pairs(s) {
+        attrs.insert(name, value);
+    }
+}
+
+fn name_value_pairs(s: &str) -> AttrIter {
+    AttrIter { s }
+}
+
+struct AttrIter<'a> {
+    s: &'a str,
+}
+
+impl<'a> Iterator for AttrIter<'a> {
+    type Item = (String, String);
+
+    fn next(&mut self) -> Option<(String, String)> {
+        let equal = self.s.find('=')?;
+        let open_quote = self.s[equal..].find('"')?;
```

Comment:
I don't think this is something the Perl version supports?

Reply:
And the way this code is written, it looks like it _requires_ there to be quotes? Or am I missing something?

Reply:
In fact, I think the Perl code specifically just has the key be everything up until the first `=`, and value be everything that follows until the next `\t` or the end of the string?

Reply:
You're right that the Perl version doesn't actually parse the value for `g_extra` and `a_extra`. It does no checks to see if they are valid attributes and just includes the string as is in either the `g` or `a` tag. But we use quick_xml so we need to parse out the key-value pairs of the extra attributes so we can add them with `extend_attributes`.

Not having quotes on the extra attributes will cause issues in the Perl version since the attribute values in the SVG tags need to be quoted. That's why I require them in this version. 

I don't like the file format they chose because it's inconsistent. If you add quotes to the other attribute values, the quotes get included in the value so it's probably not what you want. But for extra attributes you need to have quotes or they won't be there on the tag. 

For example: if we want the `class` to be "inferno" we have to do `class=inferno` which becomes `class="inferno"` on the tag. But if we wan't to change the `stroke` of the `g` tag we have to do `g_extra=stroke="green"`.

Reply:
Ah, I see what you mean... Hmm, that's unfortunate indeed. Here's what I would propose: make `"` quoting optionally available for all values. Specifically, do the following for all attributes (whether extra or non-extra):

 - `split('\t')` to get each `name=value` pair
 - `splitn(2, '=')` to get name and value
 - `trim()` both `name` and `value`
 - if `value.starts_with('"') && value.ends_with('"')`, then remove both of them

This isn't perfect, since people may start doing weird stuff like `\tname="val\tue"\t` and expect it to work correctly. Or something like `name="val\"ue"`. But at least it makes the handling consistent across all attribute types: the `"`s are always optional.

Reply:
Sounds good to me.

Reply:
Except that the extra attributes still need to be parsed a little differently since they can't be tab separated.

Reply:
Ohh, I see, `g_extra` is actually of the form
```
g_extra=foo="bar" z="baz"
```
That's awkward.. Yeah, then I guess I would say that all attribute values should be allowed to be (but not required to be) `"`-enclosed. For `g_extra`, that would mean that these are all equivalent:
```
g_extra=foo=bar z="baz"
g_extra=foo="bar" z="baz"
g_extra="foo="bar" z="baz""
g_extra="foo=bar z="baz""
```
What an awful syntax :sweat_smile:

Reply:
So I can't split on tabs for those, but I can't simply split on whitespace either because the values themselves could have spaces. What should we do about something like `g_extra=a= b  c d e = f`? Should this become `a="b  c d" e="f"`?

Reply:
It's tricky, because in HTML, that would parse as `a="b" c d e="f"`, since HTML allows "[boolean attributes](https://w3c.github.io/html/infrastructure.html#sec-boolean-attributes)" whose value is implicitly set to the name of the attribute itself. People might expect the same here, even though XML [does not allow](https://www.w3.org/TR/2006/REC-xml11-20060816/#sec-starttags) that same construct.

I think what we should do is as follows: values are *either* `"`-enclosed, or the next non-whitespace sequence of characters. If a token does not match that, it is ignored and a warning is issued. So, parsing `g_extra=a= b  c d e = f` should give:
```
WARN extra attribute c has no value (did you mean to quote the value?)
WARN extra attribute d has no value (did you mean to quote the value?)
```
and the attributes `a="b"` and `e="f"`.

Reply:
That makes sense.

Reply:
I think I've got it working with your described behavior in the latest commit.

Reply:
Yup, looks good to me!